packages/workflow/src/workflow.ts

```diff
@@ -1356,21 +1372,98 @@ export function upsertSearchAttributes(searchAttributes: SearchAttributes): void
     throw new Error('searchAttributes must be a non-null SearchAttributes');
   }
 
-  activator.pushCommand({
-    upsertWorkflowSearchAttributes: {
-      searchAttributes: mapToPayloads(searchAttributePayloadConverter, searchAttributes),
-    },
-  });
+  if (Array.isArray(searchAttributes)) {
+    // Typed search attributes
+    activator.pushCommand({
+      upsertWorkflowSearchAttributes: {
+        searchAttributes: typedMapToPayloads<string, TypedSearchAttributeUpdateValue>(
+          typedSearchAttributePayloadConverter,
+          Object.fromEntries(searchAttributes.map(([k, v]) => [k.name, v]))
+        ),
+      },
+    });
 
-  activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
-    return {
-      ...info,
-      searchAttributes: {
-        ...info.searchAttributes,
-        ...searchAttributes,
+    activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
+      const newSearchAttributes: SearchAttributes = {};
+      for (const pair of searchAttributes) {
+        const [k, v] = pair;
+        if (v === null) {
+          // If the value is null (i.e. delete), set the value to an empty list.
+          newSearchAttributes[k.name] = [];
+        } else {
+          const typedValue = v[1];
+          newSearchAttributes[k.name] = Array.isArray(typedValue) ? typedValue : ([typedValue] as SearchAttributeValue);
+        }
+      }
+      return {
+        ...info,
+        searchAttributes: {
+          ...info.searchAttributes,
+          ...newSearchAttributes,
+        },
+        typedSearchAttributes: info.typedSearchAttributes.updateAttributes(searchAttributes),
+      };
+    });
+  } else {
+    // Legacy search attributes
+    activator.pushCommand({
+      upsertWorkflowSearchAttributes: {
+        searchAttributes: mapToPayloads(searchAttributePayloadConverter, searchAttributes),
       },
-    };
-  });
+    });
+
+    activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
+      // Use updateSearchAttributes to get a deep copy of the current state.
+      let typedSearchAttributes = info.typedSearchAttributes.updateAttributes([]);
+      for (const [k, v] of Object.entries(searchAttributes)) {
+        const typedKey = TypedSearchAttributes.getKeyFromUntyped(k, v);
```

Comment:
Seems like the code here (`getKeyFromUntyped` and below) wants to be permissive, kind of allowing going through with `v` being a single primitive, rather than an array, but then you'll get until line 1442 and fail on `newValue = v[0]`. You may also get weird behaviors if `v` is set to the single value `false` or `0`, because of the `!v` check at line 1427.

Technically speaking, the former API only allowed `v` to be an array or undefined, so let's just _throw_ early if `v` anything else than that.

Reply:
`v` is recognized as `SearchAttributeValueOrReadonly` within the loop, which does imply array or undefined.

I don't get any compiler issues here given that typing, but I've added an error check just in case if we want to be extra safe.